### PR TITLE
feat(wasm): ABCI hooks for abic.BeginBlock and abci.EndBlock as wasm SudoMsg calls

### PR DESCRIPTION
**Pull request by Unique-Divine**
# feat(wasm): Wasm ABCI block hooks with sudo-governed registry

Adds begin-block and end-block Wasm hooks so a sudo-configured registry contract can return a dispatch plan of privileged `Sudo` calls each block. The chain validates plans, executes targets in isolated cache contexts, and emits structured events for indexers—enabling protocol automation without mempool exposure.

## Key Changes

1. **Sudo-governed registry address** — `x/sudo` stores `wasm_block_hooks_contract` in genesis/state and exposes `edit_wasm_block_hooks_contract` via `MsgEditSudoers` (root-only). `x/wasm` reads the registry through a narrow `WasmBlockHooksContractSource` interface wired in `app/keepers.go`.

1. **ABCI hook planner and dispatcher** — `BeginBlockWasmHooks` / `EndBlockWasmHooks` query the registry with `{"begin_block_plan":{}}` or `{"end_block_plan":{}}`, decode `WasmSudoMsgCall` items, enforce v1 guardrails (max 32 dispatches, 16 KiB JSON object payloads, 32-byte contract addresses), and run each valid target via `Keeper.Sudo` behind `CacheContext` so failures roll back without blocking siblings.

1. **Event taxonomy for observability** — Plan/query failures emit `wasm_block_hook/<hook>_failed`; completed hooks emit `wasm_block_hook/summary` with `total` and an optional JSON `failures` blob (per-item idx, contract, reason, and original `wasm_sudo_msg`). Replaces the earlier per-dispatch failure event shape.

1. **Integration tests and fixture contract** — `keeper/abci_hooks_test.go` exercises the full flow with `wasm_block_hooks_tester.wasm` (happy path, empty plan, registry query errors, mixed validation failures, and target rollback). `TestValidateWasmSudoMsgCalls` covers validation edge cases.

1. **Test and docs housekeeping** — Consolidates `x/wasm/testdata/` at module scope, renames keeper tests to `*_unit_test.go` / `*_test.go` conventions, merges `msg_server_integration_test.go` into `msg_server_test.go`, and refreshes `x/wasm/README.md` (renames `Governance.md` → `01-gov-txs.md`, `IBC.md` → `02-ibc.md`).

## Appendix

**Trust model**

Only the sudo root can set the registry. The registry can request privileged `Sudo` calls to arbitrary target contracts each block (up to 32 per hook). Operators should treat the registry contract as chain-critical infrastructure.

**Known v1 limitations / follow-ups**

- ABCI hooks run on deliver context with an effectively unbounded gas meter; consider a dedicated hook gas budget before mainnet reliance on heavy dispatches.
- Multi-block ABCI tests (`MoveToNextBlock` + `MsgEditSudoers` tx path) and localnet smoke coverage are natural next steps for operational confidence.

### CODE REVIEW

File: x/sudo/keeper/querier.go [hunk 1]
```diff
@@ -25,6 +25,16 @@ func (k Keeper) QuerySudoers(
...
 
+// GetWasmBlockHooksContract returns the configured Wasm block-hooks contract
+// address and whether the address is present.
+func (k Keeper) GetWasmBlockHooksContract(ctx sdk.Context) (sdk.AccAddress, bool) {
+	contract := k.WasmBlockHooksContract.GetOr(ctx, "")
+	if contract == "" {
+		return nil, false
+	}
+	return sdk.MustAccAddressFromBech32(contract), true
+}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/sudo/keeper/querier.go:40]: Fixed in bbfb199bb: `GetWasmBlockHooksContract` now uses `sdk.AccAddressFromBech32` and returns `nil, false` on invalid stored addresses instead of panicking. Added `TestGetWasmBlockHooksContract_InvalidStoredAddress` and wasm hook coverage for invalid stored registry.

File: x/wasm/keeper/abci_hooks.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	hookKind wasmBlockHookKind,
+) (wasmSudoMsgCalls []WasmSudoMsgCall, enabled bool, err error) {
+	if k.wasmBlockHooksContractSource == nil {
+		return nil, false, nil
+	}
+
+	registryAddr, configured := k.wasmBlockHooksContractSource.GetWasmBlockHooksContract(ctx)
+	if !configured {
+		return nil, false, nil
+	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/wasm/keeper/abci_hooks.go:142]: Fixed in bbfb199bb: `queryWasmSudoMsgCalls` now treats `!configured || registryAddr == nil` as feature-off and returns no calls.

File: x/sudo/keeper/keeper.go [hunk 3]
```diff
@@ -137,6 +144,40 @@ func (k Keeper) RemoveContracts(
...
+	if err != nil {
+		return nil, err
+	}
+	k.WasmBlockHooksContract.Set(ctx, contractAddr)
+
+	msgResp = new(sudo.MsgEditSudoersResponse)
+	return msgResp, ctx.EventManager().EmitTypedEvent(&sudo.EventUpdateSudoers{
+		Sudoers: pbSudoers,
+		Action:  msg.Action,
+	})
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/sudo/keeper/keeper.go:182]: Fixed in bbfb199bb: `EditWasmBlockHooksContract` now emits a plain SDK `wasm_block_hooks_contract_update` event with the new registry address (empty string when cleared), in addition to the existing typed `EventUpdateSudoers`. Covered in `msg_server_test.go`.

File: x/wasm/keeper/abci_hooks.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+		_, err = k.Sudo(cacheCtx, validated.ContractAddr, validated.Msg)
+		if err != nil {
+			failures = appendWasmBlockHookDispatchFailure(failures, idx, wasmSudoMsg, err)
+			continue
+		}
+
+		writeCacheCtx()
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/wasm/keeper/abci_hooks.go:112]: Thanks for flagging this. I traced the call path end to end before deciding whether to add panic recovery around `k.Sudo` in function `runWasmBlockHook`. I don't think a code change is warranted here.

## What actually happens on contract failure

ABCI hooks call function `Keeper.Sudo`, the same entrypoint used by other privileged wasm sudo paths (for example message `MsgSudoContract`). That function does not use `recover`. Instead, it relies on wasmvm to return VM failures as Go errors:

1. function `runWasmBlockHook` calls `k.Sudo(cacheCtx, ...)`.
2. function `Keeper.Sudo` calls `k.wasmVM.Sudo(...)`.
3. If the contract fails — including an explicit Rust panic inside the Wasm guest — wasmvm returns `execErr`.
4. The keeper wraps that as error `ErrExecuteFailed` and returns it. No Go panic is raised.

This behavior is already covered by keeper tests. For example, function `TestExecuteWithPanic` sends message `{"panic":{}}` to a contract and asserts the node gets error `ErrExecuteFailed` with text like `panicked at 'This page intentionally faulted'`, not a process crash. The wasm-block-hooks-tester fixture uses `anyhow::bail!` for `fail_before_write` and `fail_after_write`, which follow the same error-return path.

Our ABCI hook tests exercise that path directly: a failing target sudo call is recorded in the hook summary failures payload, state written inside the failing call is rolled back via `CacheContext`, and later valid calls still run.

## Rust/Wasm panics vs Go panics

The review comment mixes two different failure modes:

| Failure mode | What happens | ABCI hook behavior |
|---|---|---|
| Rust/Wasm contract panic or `ContractError` | wasmvm returns an error; function `Keeper.Sudo` returns `ErrExecuteFailed` | Caught by the existing `if err != nil` branch; block continues |
| Go panic inside keeper/submessage dispatch | Propagates up the stack | Not what normal contract failures produce |

For governed hook automation, the first row is the operational case we care about. That path is already guarded without `recover`.

## SDK out-of-gas panics during BeginBlock/EndBlock

Function `consumeRuntimeGas` can panic with `sdk.ErrorOutOfGas`, but only when `ctx.GasMeter()` reports out of gas. During BeginBlock and EndBlock, the deliver-state context uses an infinite tx gas meter (see function `sdk.NewContext`). Wasm sudo execution charges against `ctx.GasMeter()`, not `ctx.BlockGasMeter()`. Under normal block processing, an SDK OOG panic from `consumeRuntimeGas` during hooks is therefore not expected.

Submessages use a separate recovery path in function `dispatchMsgWithGasLimit`, which catches OOG panics from limited sub-calls and converts them to errors. Non-OOG panics there are re-raised, which matches existing wasm keeper behavior outside ABCI hooks.

## Why I am not adding `recover` here

1. **Redundant for contract failures.** wasmvm already converts guest failures to returned errors. Wrapping `k.Sudo` in `recover` would mostly duplicate behavior the keeper already provides.
2. **Consistent with other sudo entrypoints.** Function `Keeper.Sudo`, message `MsgSudoContract`, and other direct sudo callers do not wrap execution in `recover`. ABCI hooks should follow the same contract-failure model.
3. **Cosmos SDK does not recover BeginBlock/EndBlock module hooks.** Unexpected Go panics during block hooks are treated as serious failures. Masking them at the hook layer could hide keeper bugs that should halt rather than be logged as a single failed dispatch item.
4. **Tests already validate the intended failure semantics.** Invalid items, registry query errors, target sudo failures, rollback, event isolation, and continuation are covered in `abci_hooks_test.go`.

## Conclusion

Contract-side failures during ABCI wasm hooks are already handled safely through the normal wasmvm error path plus per-call `CacheContext` isolation. Adding panic recovery around `k.Sudo` would not materially improve safety for expected contract failure modes, and would diverge from how the rest of the wasm keeper treats sudo execution.

No code change planned for this comment.
